client/dashboard/app/i18n.tsx

```diff
@@ -0,0 +1,55 @@
+import { defaultI18n } from '@wordpress/i18n';
+import { I18nProvider as WPI18nProvider } from '@wordpress/react-i18n';
+import { useEffect, useRef, useState, type PropsWithChildren } from 'react';
+import { useAuth } from './auth';
+
+async function fetchLocaleData( language: string ) {
+	if ( language === 'en' ) {
+		return [ language, undefined ];
+	}
+
+	try {
+		const response = await fetch(
+			`https://widgets.wp.com/languages/calypso/${ language }-v1.1.json`
+		);
+
+		return [ language, await response.json() ];
+	} catch {
+		// Fall back to `en` when fetching the language data fails. Without this
+		// the i18n provider would be stuck forever in a non-loaded state.
+		return [ 'en', undefined ];
+	}
+}
+
+export function I18nProvider( { children }: PropsWithChildren ) {
+	const [ loadedLocale, setLoadedLocale ] = useState< string | null >( null );
+	const loadingLocaleRef = useRef< string | null >( null );
+	const { user } = useAuth();
+
+	const i18n = defaultI18n;
+	const language = user.locale_variant || user.language || 'en';
+
+	useEffect( () => {
+		loadingLocaleRef.current = language;
+
+		fetchLocaleData( language ).then( ( [ realLanguage, data ] ) => {
+			// Activate the data only if no other language switch has run in the meantime
+			if ( loadingLocaleRef.current !== language ) {
+				return;
+			}
+
+			i18n.resetLocaleData( data );
+			// `realLanguage` can be different from `language` when loading language data fails
+			// and it falls back to `en`.
+			setLoadedLocale( realLanguage );
+		} );
+	}, [ i18n, language ] );
+
+	// Render the sub-tree only after the initial locale data are loaded. We don't want a
+	// flash of `en` content on the initial render that's updated a moment later.
+	if ( loadedLocale === null ) {
+		return null;
+	}
```

Comment:
I think this makes sense as a first pass, but it feels kinda thrashy to show nothing until the network request resolves. I also wonder if it could create issues if we ever did server-rendering. Or maybe we should have children be a function so that we don't spend any effort creating the element tree until we need it. Worst-case: some kind of loader state.

I think ideally with your "bootstrap" next steps idea, this could all be synchronous though.

Reply:
It would be great if we can simply load the locale data from the server side. I remember we have already had this mechanism and data is stored in somewhere like `window.i18nLanguageManifest`.

Reply:
I didn't want to do the initial render with `en` locale and only later switch to the real locale. Better to wait until we have all the data. It's a tradeoff of course.

A similar wait is done in the `AuthProvider` component. It won't render the subtree until the user object is loaded from the `/me` endpoint.

If we want to be ready for SSR, the first thing we should probably do is to rewrite both `AuthProvider` and `I18nProvider` to use Suspense. I guess the TanStack Router that we use here already suspends a lot! Then the components wouldn't render `null`, but would simply suspend and continue rendering only when the needed data are loaded.

Reply:
> It would be great if we can simply load the locale data from the server side.

Yes, that's the next thing I'd like to add support for, because it's useful for the Hosting dashboard.